src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+        m_src_mirror_image.state != cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
+      ldout(m_cct, 10) << "mirroring is not enabled for image" << dendl;
+      return 0;
+    }
+  }
+
+  C_SaferCond ctx;
+  auto req = mirror::EnableRequest<I>::create(m_dst_io_ctx, m_dst_image_id, "",
+                                              m_image_ctx->op_work_queue, &ctx);
+  req->send();
+  int r = ctx.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to enable mirroring: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::reopen_image() {
+  ldout(m_cct, 10) << dendl;
+
+  C_SaferCond on_close;
+  m_image_ctx->state->close(&on_close);
+  int r = on_close.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to close image: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  m_image_ctx->clear_state();
+
+  m_image_ctx->cct = m_cct;
+
+  ThreadPool *thread_pool;
+  m_image_ctx->get_thread_pool_instance(m_cct, &thread_pool,
+                                        &m_image_ctx->op_work_queue);
+
+  assert(m_image_ctx->io_work_queue == nullptr);
+  m_image_ctx->io_work_queue = new io::ImageRequestWQ<I>(
+      m_image_ctx, "librbd::io_work_queue",
+      m_cct->_conf->get_val<int64_t>("rbd_op_thread_timeout"), thread_pool);
+
+  assert(m_image_ctx->state == nullptr);
+  m_image_ctx->state = new ImageState<I>(m_image_ctx);
+
+  m_image_ctx->old_format = false;
+  m_image_ctx->id = m_dst_image_id;
+  m_image_ctx->md_ctx.dup(m_dst_io_ctx);
+  m_image_ctx->data_ctx.dup(m_dst_io_ctx);
+
+  C_SaferCond on_open;
+  m_image_ctx->state->open(false, &on_open);
+  r = on_open.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed opening newly created image: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::remove_src_image() {
+  ldout(m_cct, 10) << dendl;
+
+  m_src_migration_spec.state = cls::rbd::MIGRATION_STATE_COMPLETE;
+  int r = cls_client::migration_set(&m_src_io_ctx, m_src_header_oid,
+                                    m_src_migration_spec);
+  if (r < 0) {
+    lderr(m_cct) << "failed updating migration header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  r = purge_src_image_snaps();
+  if (r < 0) {
+    lderr(m_cct) << "failed purging snapshots: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  if (m_src_old_format) {
+    ldout(m_cct, 10) << "removing " << m_src_image_name << dendl;
+
+    ThreadPool *thread_pool;
+    ContextWQ *op_work_queue;
+    ImageCtx::get_thread_pool_instance(m_cct, &thread_pool, &op_work_queue);
+    librbd::NoOpProgressContext prog_ctx;
+    C_SaferCond on_remove;
+    auto req = librbd::image::RemoveRequest<I>::create(
+      m_src_io_ctx, m_src_image_name, "", false, false, prog_ctx,
+      op_work_queue, &on_remove);
+    req->get_image_ctx()->old_format = true; // XXXMG
```

Comment:
... could add a new constructor for ```RemoveImage``` that takes an already opened image. Then you could keep the source image context open for the full migration and re-use here.